modules/platforms/dotnet/Apache.Ignite.Core.Tests/Cache/CacheAbstractTransactionalTest.cs

```diff
@@ -862,6 +862,50 @@ public void TestTransactionScopeAllOperations()
             }
         }
 
+        /// <summary>
+        /// Tests that read operations lock keys in Serializable mode.
+        /// </summary>
+        [Test]
+        public void TestTransactionScopeWithSerializableIsolationLocksKeysOnRead()
+        {
+            Action<Func<ICache<int, int>, int, int>>
+                test = TestTransactionScopeWithSerializableIsolationLocksKeysOnRead;
+
+            test((cache, key) => cache[key]);
+            test((cache, key) => cache.Get(key));
+            test((cache, key) => cache.GetAsync(key).Result);
+            test((cache, key) => { int val; return cache.TryGet(key, out val) ? val : 0; });
+            test((cache, key) => cache.TryGetAsync(key).Result.Value);
+            test((cache, key) => cache.GetAll(new[] {key}).Single().Value);
+            test((cache, key) => cache.GetAllAsync(new[] {key}).Result.Single().Value);
+        }
+
+        /// <summary>
+        /// Tests that read operations lock keys in Serializable mode.
+        /// </summary>
+        private void TestTransactionScopeWithSerializableIsolationLocksKeysOnRead(
+            Func<ICache<int, int>, int, int> readOp)
+        {
+            var cache = Cache();
+            cache.Put(1, 1);
+
+            var options = new TransactionOptions {IsolationLevel = IsolationLevel.Serializable};
+
+            using (var scope = new TransactionScope(TransactionScopeOption.Required, options))
+            {
+                Assert.AreEqual(1, readOp(cache, 1));
+
+                var taskFinished = Task.Factory.StartNew(() => cache.Put(1, 2)).Wait(TimeSpan.FromSeconds(1));
+
```

Comment:
To be honest I prefer check the optimistic serializable transaction for throwing exception (as defined in [documetation](https://apacheignite.readme.io/docs/concurrency-modes-and-isolation-levels#optimistic-transactions)) on commit it is seems to me to be more predictable and no need for magic numbers and waits.

Reply:
Or just check that transaction is started.

Reply:
Or just check that transaction is started.

Reply:
Good point, I've added current Tx check, and replaced time-based wait with a wait handle.
But I'd like to keep the logic that checks whether keys are locked.

Optimistic transactions are a different matter, we can have a separate test for them.

Reply:
I've added more tests for Optimistic transactions, please have a look.